base/iobuffer.jl

```diff
@@ -536,38 +739,85 @@ _unsafe_take!(io::IOBuffer) =
         io.size - io.offset)
 
 function write(to::IO, from::GenericIOBuffer)
-    written::Int = bytesavailable(from)
+    available = bytesavailable(from)
+    # If they're the same buffer, we need to special case it since the buffer
+    # is being mutated twice in the write call.
     if to === from
-        from.ptr = from.size + 1
+        if from.append
+            # If from.append, we copy data to from.size + 1.
+            # Note that this ensureroom might switch the data buffer, or update
+            # the fields like from.size
+            ensureroom(from, available)
+            data = from.data
+            size = from.size
+            existing_space = lastindex(data) - min(from.maxsize, size)
+            to_write = min(existing_space, available)
+            iszero(to_write) && return 0
+            GC.@preserve from unsafe_copyto!(data, size + 1, data, from.ptr, to_write)
+            from.size = size + to_write
+            from.ptr += to_write
```

Comment:
even if the write failed, in theory we still read that data
```suggestion
            from.ptr = from.size + 1
```

(I'd also support just making this whole case an error, since it should really be an infinite loop per the generic definition of reading until eof)

Reply:
What you suggest here is inconsistent with how `write(::IOBuffer, ::IOBuffer)` currently works when it's to distinct buffers. Consider:
```julia
julia> buf = IOBuffer(;maxsize=3);
       buf2 = IOBuffer(collect(b"abcdefgh"));
       write(buf, buf2)
       println(String(read(buf2)))
defgh
```
If I understand correctly, you suggest that in this example, `buf2` should have been read until completion, and should now be empty?

I see that the existing behaviour is different from the generic definition:
```julia
julia> buf = IOBuffer(;maxsize=3);
       buf2 = IOBuffer(collect(b"abcdefgh"));
       @invoke write(buf::IO, buf2::IO)
       eof(buf2)
true
```
So I'll change the behaviour in the following way:
* Error when `to === from`, as you suggest
* Make sure that, when `to !== from`, all of `from` is read, even if `to` is full.

Reply:
Yes, that is correct, I guess I forgot to update the read pointer on `from` when `from !== to`. Good catch